### PR TITLE
Fix issue #16451: Don't stop an outer subquery early due to inner LIM…

### PR DESCRIPTION
**Pull request by hkernbach**
### Scope & Purpose

3.8 Backport of:
* https://github.com/arangodb/arangodb/pull/17644

Please see all details there. 

- [x] :hankey: Bugfix

### CODE REVIEW

File: arangod/Aql/AqlItemBlockInputMatrix.cpp [hunk 1]
```diff
@@ -64,26 +57,20 @@ AqlItemBlockInputRange& AqlItemBlockInputMatrix::getInputRange() {
...
-  TRI_ASSERT(_aqlItemMatrix == nullptr);
-  return _block;
-}
-
-std::pair<ExecutorState, AqlItemMatrix const*> AqlItemBlockInputMatrix::getMatrix() noexcept {
+std::pair<ExecutorState, AqlItemMatrix const*>
+AqlItemBlockInputMatrix::getMatrix() noexcept {
   TRI_ASSERT(_aqlItemMatrix != nullptr);
-  TRI_ASSERT(_block == nullptr);
-  TRI_ASSERT(!_shadowRow.isInitialized());
```
Comment (goedderz) [inline arangod/Aql/AqlItemBlockInputMatrix.cpp:86]: Why did you remove these assertions?

File: arangod/Aql/AqlItemBlockInputMatrix.cpp [hunk 2]
```diff
@@ -41,16 +41,9 @@ AqlItemBlockInputMatrix::AqlItemBlockInputMatrix(ExecutorState state)
...
-    : _block{block}, _aqlItemMatrix{nullptr} {
-  TRI_ASSERT(_aqlItemMatrix == nullptr);
-  TRI_ASSERT(!hasDataRow());
-}
-
-AqlItemBlockInputMatrix::AqlItemBlockInputMatrix(ExecutorState state, AqlItemMatrix* aqlItemMatrix)
+AqlItemBlockInputMatrix::AqlItemBlockInputMatrix(ExecutorState state,
+                                                 AqlItemMatrix* aqlItemMatrix)
     : _finalState{state}, _aqlItemMatrix{aqlItemMatrix} {
-  TRI_ASSERT(_block == nullptr);
```
Comment (goedderz) [inline arangod/Aql/AqlItemBlockInputMatrix.cpp:53]: Why did you remove this assertion?

File: arangod/Aql/AqlItemBlockInputRange.cpp [hunk 3]
```diff
@@ -183,7 +189,10 @@ size_t AqlItemBlockInputRange::skipAllShadowRowsOfDepth(size_t depth) {
 }
 
 
-template <AqlItemBlockInputRange::LookAhead doPeek, AqlItemBlockInputRange::RowType type>
+template size_t AqlItemBlockInputRange::skipAllShadowRowsOfDepth<-1>(
+    size_t depth);
+template size_t AqlItemBlockInputRange::skipAllShadowRowsOfDepth<0>(
+    size_t depth);template <AqlItemBlockInputRange::LookAhead doPeek, AqlItemBlockInputRange::RowType type>
```
Comment (goedderz) [inline arangod/Aql/AqlItemBlockInputRange.cpp:195]:
```suggestion
template size_t AqlItemBlockInputRange::skipAllShadowRowsOfDepth<0>(
    size_t depth);

template <AqlItemBlockInputRange::LookAhead doPeek, AqlItemBlockInputRange::RowType type>
```

File: arangod/Aql/AqlItemBlockInputRange.cpp [hunk 4]
```diff
@@ -158,22 +158,28 @@ size_t AqlItemBlockInputRange::countAndSkipAllRemainingDataRows() {
   return skipped;
 }
 
+template<int depthOffset>
 size_t AqlItemBlockInputRange::skipAllShadowRowsOfDepth(size_t depth) {
+  // Note that depthOffset == -1 iff we're at an SQS node.
```
Comment (goedderz) [inline arangod/Aql/AqlItemBlockInputRange.cpp:163]:
```suggestion
template<int depthOffset>
size_t AqlItemBlockInputRange::skipAllShadowRowsOfDepth(size_t depth) {
  static_assert(depthOffset == 0 || depthOffset == -1);
  // Note that depthOffset == -1 iff we're at an SQS node.
```

Comment (mchacki) [reply arangod/Aql/AqlItemBlockInputRange.cpp:163]: :+1:

**Review by mchacki — CHANGES_REQUESTED** [review]
Same as @goedderz already said.

In general LGTM, I would like to get the assertions back as well.

File: arangod/Aql/AqlItemBlockInputMatrix.cpp (hunk 1 above)
Comment (hkernbach) [reply arangod/Aql/AqlItemBlockInputMatrix.cpp:86]: Ups. "By accident". This might happen during the cherry-pick. Will revert this.

Comment (hkernbach) [reply arangod/Aql/AqlItemBlockInputMatrix.cpp]: re-added asserts.

File: arangod/Aql/AqlItemBlockInputMatrix.cpp (hunk 2 above)
Comment (hkernbach) [reply arangod/Aql/AqlItemBlockInputMatrix.cpp:53]: re-added asserts.

**Review by goedderz — APPROVED** [review]

**Review by mchacki — APPROVED** [review]
LGTM
